src/ntfs/write.rs

```diff
@@ -239,181 +215,283 @@ struct Fares<'a> {
-            after_change: format!("network=network:{}", ticket_use_perimeter.object_id),
-            ..Default::default()
-        });
-        for connected_perimeter in fares
-            .ticket_use_perimeters
+        // Now the ticket for our ticket_use_id.
+        //  there cannot exists two Ticket with the same ticket_id in fares.tickets
+        //  thus it is sufficient to check if one ticket exists with the requested ticket_id
+        let ticket = fares.tickets.get(&ticket_use.ticket_id).ok_or_else(|| {
+            format_err!(
+                "The ticket_id {:?} was not found in tickets.txt",
+                ticket_use.ticket_id
+            )
+        })?;
+
+        //We have everything, so let's fill the fare v1 data !
+
+        //first  prices_v1
+        // we find all prices with id ticket.id
+        // and for each we create a price_v1 with id ticket_use_id (as ticket_use_id of fare v2 plays the role of ticket_id in fare v1)
+        for price in fares
+            .ticket_prices
             .values()
-            .filter(|p| p.object_type == ObjectType::Network)
-            .filter(|p| p.perimeter_action == PerimeterAction::Included)
-            .filter(|p| p.ticket_use_id == ticket_use_perimeter.ticket_use_id)
-            .filter(|p| p.object_id != ticket_use_perimeter.object_id)
+            .filter(|&ticket_price| ticket_price.ticket_id == ticket.id)
         {
-            fares_v1.insert(FareV1 {
-                before_change: format!("network=network:{}", ticket_use_perimeter.object_id),
-                after_change: format!("network=network:{}", connected_perimeter.object_id),
-                ..Default::default()
-            });
+            // For now we restrict to EUR only.
+            // There is several reasons to that :
+            // - fare v1 needs prices to be all in the same currency
+            // - if we want to support several currencies, we would need to have access to currency exchange rates here
+            //   and it's unclear how to provide this information (which evolves over time)
+            if price.currency != "EUR" {
+                warn!(
+                    "The price {:?} is ignored as it has an unsupported currency : {}. \
+                     Only EUR currency supported in conversion from fare v2 to fare v1.",
+                    price, price.currency
+                );
+                continue;
+            }
+            let price_v1 = build_price_v1(&ticket_use.id, ticket, price)?;
+            prices_v1.insert(price_v1);
+        }
+
+        //now fares_v1
+        {
+            let states = included_networks
+                .iter()
+                .map(|network| format!("network=network:{}", network))
+                .chain(
+                    included_lines
+                        .iter()
+                        .map(|line| format!("line=line:{}", line)),
+                );
```

Comment:
```suggestion
                )
                .collect::<Vec<String>>();
```
if you collect here you can avoid an extra clone below (and just using a reference).
```&states``` instead of ```states.clone()```

Reply:
but then it will allocate a vector and all the strings inside it, whereas for now it just clones the iterator, right ?

Reply:
@pbench Right. But you can avoid multiple clones afterwards (and therefore memory allocation).

Reply:
allright ! updated